__Домашняя работа 7.__

In [1]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris

__Задание 1.__ К алгоритму kNN, реализованному на уроке, реализовать добавление весов для соседей по любому из показанных на уроке принципов.

In [2]:
def e_metrics(x1, x2):
    
    distance = np.sum(np.square(x1 - x2))

    return np.sqrt(distance)

In [3]:
def knn(x_train, y_train, x_test, k, q):
    answers = []
    for x in x_test:
        test_distances = []
            
        for i in range(len(x_train)):
            
            # расчет расстояния от классифицируемого объекта до
            # объекта обучающей выборки
            distance = e_metrics(x, x_train[i])
            
            # Записываем в список значение расстояния и ответа на объекте обучающей выборки
            test_distances.append((distance, y_train[i]))
        
        # создаем словарь со всеми возможными классами
        classes = {class_item: 0 for class_item in set(y_train)}
        
        # Сортируем список и среди первых k элементов подсчитаем частоту появления разных классов
        for d in sorted(test_distances)[0:k]:
            #ИЗМЕНЕНИЯ:
            classes[d[1]] += q**d[0]       #1
        # Записываем в список ответов наиболее часто встречающийся класс
        answers.append(sorted(classes, key=classes.get)[-1])
        
    return answers

In [4]:
def accuracy(pred, y):
    return (sum(pred == y) / len(y))

In [5]:
X, y = load_iris(return_X_y=True)

# Для наглядности возьмем только первые два признака (всего в датасете их 4)
X = X[:, :2]

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size=0.2,
                                                    random_state=1,
                                                    stratify=y)

In [7]:
for q in np.linspace(0, 1, 10):
    ans = []
    for i in range(1,100):
        y_pred = knn(X_train, y_train, X_test, i, q)
        ans.append(accuracy(y_pred, y_test))
    f = {x: ans[x-1] for x in range(1,100)}
    k = sorted(f, key=f.get)[-1]
    print(f'Точность алгоритма при k = {k}, q = {q:.3f}: {f[k]:.3f}')

Точность алгоритма при k = 99, q = 0.000: 0.533
Точность алгоритма при k = 36, q = 0.111: 0.767
Точность алгоритма при k = 37, q = 0.222: 0.767
Точность алгоритма при k = 37, q = 0.333: 0.767
Точность алгоритма при k = 37, q = 0.444: 0.767
Точность алгоритма при k = 37, q = 0.556: 0.767
Точность алгоритма при k = 37, q = 0.667: 0.767
Точность алгоритма при k = 37, q = 0.778: 0.767
Точность алгоритма при k = 37, q = 0.889: 0.767
Точность алгоритма при k = 38, q = 1.000: 0.767


Вывод: Добавление весов от расстояния никак не влияет на точность предсказания. Больше 0.767 поднять точность не удалось